File: YKOCBaseClass/Classes/YKNavigationController.m
//
//  YKNavigationController.m
//  YYKPodLib_Sec
//
//  Created by edward on 2020/10/10.
//  Copyright © 2020 edward. All rights reserved.
//

#import "YKNavigationController.h"

@interface YKNavigationController ()<UINavigationControllerDelegate>


@property (nonatomic,strong) id popDelegate;
/** 返回按钮 */
@property (nonatomic, strong) UIButton *backBtn;
/** 返回按钮 */
@property (nonatomic, strong) UIBarButtonItem *backBtnItem;
/** 最后push的controller的类名 */
@property (nonatomic, copy) NSString *latestPushViewControllerName;
@end

@implementation YKNavigationController

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.modalPresentationStyle = UIModalPresentationFullScreen;
    }
    return self;
}

- (instancetype)initWithRootViewController:(UIViewController *)rootViewController
{
    if (self = [super initWithRootViewController:rootViewController]) {
        self.modalPresentationStyle = UIModalPresentationFullScreen;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.delegate = self;
}

- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated
{
    if (!viewController) return;
    if (self.childViewControllers.count > 0) {
        UIButton *backBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        
        UIImage *backImage = [UIImage new];
        if (@available(iOS 13.0, *)) {
            if (self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark) {
                backImage = [self bc_imageNamed:@"ic_bc_back_while"];
            }else
            {
                backImage = [self bc_imageNamed:@"ic_bc_back_black"];
            }
        } else {
            backImage = [self bc_imageNamed:@"ic_bc_back_black"];
        }
        backImage = [backImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        [backBtn setImage:backImage forState:UIControlStateNormal];
        [backBtn setImage:backImage forState:UIControlStateHighlighted];
        if (self.backButtonColor) {
            [backBtn setTintColor:self.backButtonColor];
        } else {
            [backBtn setTintColor:UIColor.blackColor];
        }
        backBtn.titleLabel.font = [UIFont systemFontOfSize:17];
        [backBtn setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        
        CGFloat ipad = (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) ? 2 : 1;
        CGRect frame = backBtn.frame;
        frame.size = CGSizeMake(44 * ipad, 44 * ipad);
        backBtn.frame = frame;
        backBtn.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
        [backBtn addTarget:self action:@selector(back) forControlEvents:UIControlEventTouchUpInside];
        
        self.backBtn = backBtn;
        
        viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:backBtn];
        
        self.backBtnItem = viewController.navigationItem.leftBarButtonItem;
        
        if (self.viewControllers.count == 1) {
            viewController.hidesBottomBarWhenPushed = YES;
        }
    }else{
        viewController.hidesBottomBarWhenPushed = NO;
    }
    
    [super pushViewController:viewController animated:animated];
}

- (void)back
{
    [self popViewControllerAnimated:YES];
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleDefault;
}

- (UIViewController *)childViewControllerForStatusBarStyle
{
    return self.topViewController;
}

#pragma mark - setter & getter
- (void)setBackBtnTitle:(NSString *)backBtnTitle {
    _backBtnTitle = backBtnTitle;
    if (backBtnTitle) {
        [self.backBtn setTitle:backBtnTitle forState:UIControlStateNormal];
        [self.backBtn setTitleColor:[UIColor colorWithRed:236.0/255.0 green:89.0/255.0 blue:90.0/255.0 alpha:1] forState:UIControlStateNormal];
        CGFloat ipad = (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) ? 2 : 1;
        CGRect frame = self.backBtn.frame;
        frame.size = CGSizeMake(130 * ipad, 44 * ipad);
        self.backBtn.frame = frame;
        self.backBtn.titleEdgeInsets = UIEdgeInsetsMake(0, 10, 0, 0);
        self.backBtn.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
    } else {
        [self.backBtn setTitle:backBtnTitle forState:UIControlStateNormal];
        CGFloat ipad = (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) ? 2 : 1;
        CGRect frame = self.backBtn.frame;
        frame.size = CGSizeMake(44 * ipad, 44 * ipad);
        self.backBtn.frame = frame;
        self.backBtn.titleEdgeInsets = UIEdgeInsetsMake(0, 0, 0, 0);
        self.backBtn.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
    }
}

- (void)setBackgroundColor:(UIColor *)backgroundColor
{
    _backgroundColor = backgroundColor;
    
    /**  iOS13开始改用standard
     self.navigationBar.barTintColor = backgroundColor;
     self.navigationBar.backgroundColor = backgroundColor;
     */
    if (backgroundColor) {
        if (@available(iOS 15, *)) {
            
            
            UINavigationBarAppearance *standardAppearance = self.navigationBar.standardAppearance;
            
            if (!standardAppearance) {
                standardAppearance = [[UINavigationBarAppearance alloc] init];
            }
            standardAppearance.backgroundColor = backgroundColor;
            self.navigationBar.standardAppearance = standardAppearance;
            
            UINavigationBarAppearance *scrollEdgeAppearance = self.navigationBar.scrollEdgeAppearance;
            
            if (!scrollEdgeAppearance) {
                scrollEdgeAppearance = [[UINavigationBarAppearance alloc] init];
            }
            
            scrollEdgeAppearance.backgroundColor = backgroundColor;
            self.navigationBar.scrollEdgeAppearance = scrollEdgeAppearance;
            
        } else {
            self.navigationBar.barTintColor = backgroundColor;
            self.navigationBar.backgroundColor = backgroundColor;
        }
    }
}

- (void)setTabbarBackgroundColor:(UIColor *)tabbarBackgroundColor
{
    _tabbarBackgroundColor = tabbarBackgroundColor;
    if (tabbarBackgroundColor) {
        if (@available(iOS 15, *)) {
            
            UITabBarAppearance *scrollAppearance = self.tabBarItem.scrollEdgeAppearance;
            
            if (!scrollAppearance) {
                scrollAppearance = [[UITabBarAppearance alloc] init];
            }
            scrollAppearance.backgroundColor = tabbarBackgroundColor;
            self.tabBarItem.scrollEdgeAppearance = scrollAppearance;
            
            UITabBarAppearance *standardAppearance = self.tabBarItem.standardAppearance;
            
            if (!standardAppearance) {
                standardAppearance = [[UITabBarAppearance alloc] init];
            }
            standardAppearance.backgroundColor = tabbarBackgroundColor;
            self.tabBarItem.standardAppearance = standardAppearance;
        } else {
            [[UITabBar appearance] setBarTintColor:tabbarBackgroundColor];
        }
    }
}

- (void)setTabbarNormalItemTextColor:(UIColor *)tabbarNormalItemTextColor
{
    _tabbarNormalItemTextColor = tabbarNormalItemTextColor;
    
    if (tabbarNormalItemTextColor) {
        if (@available(iOS 15, *)) {
            
            UITabBarAppearance *scrollAppearance = self.tabBarItem.scrollEdgeAppearance;
            
            if (!scrollAppearance) {
                scrollAppearance = [[UITabBarAppearance alloc] init];
            }
            
            NSMutableDictionary<NSAttributedStringKey, id> *scrollNormalAttributes = scrollAppearance.stackedLayoutAppearance.normal.titleTextAttributes.mutableCopy;
            [scrollNormalAttributes setValue:tabbarNormalItemTextColor forKey:NSForegroundColorAttributeName];
            scrollAppearance.stackedLayoutAppearance.normal.titleTextAttributes = scrollNormalAttributes.copy;
            self.tabBarItem.scrollEdgeAppearance = scrollAppearance;

            
            UITabBarAppearance *standardAppearance = self.tabBarItem.standardAppearance;
            
            if (!standardAppearance) {
                standardAppearance = [[UITabBarAppearance alloc] init];
            }
            
            NSMutableDictionary<NSAttributedStringKey, id> *standardNormalAttributes = standardAppearance.stackedLayoutAppearance.normal.titleTextAttributes.mutableCopy;
            [standardNormalAttributes setValue:tabbarNormalItemTextColor forKey:NSForegroundColorAttributeName];
            standardAppearance.stackedLayoutAppearance.normal.titleTextAttributes = standardNormalAttributes.copy;
            self.tabBarItem.standardAppearance = standardAppearance;
        } else {
            
            NSMutableDictionary<NSAttributedStringKey, id> *normalAttributes = [self.tabBarItem titleTextAttributesForState:UIControlStateNormal].mutableCopy;
            
            if (!normalAttributes) {
                normalAttributes = [NSMutableDictionary dictionary];
            }
            
            [normalAttributes setValue:tabbarNormalItemTextColor forKey:NSForegroundColorAttributeName];
            [self.tabBarItem setTitleTextAttributes:normalAttributes.copy forState:UIControlStateNormal];
            
        }
    }
}

- (void)setTabbarSelectItemTextColor:(UIColor *)tabbarSelectItemTextColor
{
    _tabbarSelectItemTextColor = tabbarSelectItemTextColor;
    if (tabbarSelectItemTextColor) {
        if (@available(iOS 15, *)) {
            UITabBarAppearance *scrollAppearance = self.tabBarItem.scrollEdgeAppearance;
            if (!scrollAppearance) {
                scrollAppearance = [[UITabBarAppearance alloc] init];
            }
            NSMutableDictionary<NSAttributedStringKey, id> *scrollSelectedAttributes = scrollAppearance.stackedLayoutAppearance.selected.titleTextAttributes.mutableCopy;
            [scrollSelectedAttributes setValue:tabbarSelectItemTextColor forKey:NSForegroundColorAttributeName];
            scrollAppearance.stackedLayoutAppearance.selected.titleTextAttributes = scrollSelectedAttributes.copy;
            self.tabBarItem.scrollEdgeAppearance = scrollAppearance;
            UITabBarAppearance *standardAppearance = self.tabBarItem.standardAppearance;
            if (!standardAppearance) {
                standardAppearance = [[UITabBarAppearance alloc] init];
            }
            NSMutableDictionary<NSAttributedStringKey, id> *standardSelectedAttributes = standardAppearance.stackedLayoutAppearance.selected.titleTextAttributes.mutableCopy;
            [standardSelectedAttributes setValue:tabbarSelectItemTextColor forKey:NSForegroundColorAttributeName];
            standardAppearance.stackedLayoutAppearance.selected.titleTextAttributes = standardSelectedAttributes.copy;
            self.tabBarItem.standardAppearance = standardAppearance;
        } else {
            NSMutableDictionary<NSAttributedStringKey, id> *selectedAttributes = [self.tabBarItem titleTextAttributesForState:UIControlStateSelected].mutableCopy;
            if (!selectedAttributes) {
                selectedAttributes = [NSMutableDictionary dictionary];
            }
            [selectedAttributes setValue:tabbarSelectItemTextColor forKey:NSForegroundColorAttributeName];
            [self.tabBarItem setTitleTextAttributes:selectedAttributes.copy forState:UIControlStateSelected];
        }
    }
}

- (void)setTabbarSelectedImage:(UIImage *)tabbarSelectedImage
{
    _tabbarSelectedImage = tabbarSelectedImage;
    self.tabBarItem.selectedImage = tabbarSelectedImage;
}

- (void)setTabbarNormalImage:(UIImage *)tabbarNormalImage
{
    _tabbarNormalImage = tabbarNormalImage;
    self.tabBarItem.image = tabbarNormalImage;
}

- (void)setTabbarTitle:(NSString *)tabbarTitle
{
    _tabbarTitle = tabbarTitle;
    self.tabBarItem.title = tabbarTitle;
}


- (void)setForegroundColor:(UIColor *)foregroundColor
{
    _foregroundColor = foregroundColor;
    NSMutableDictionary *barTitleDic = [NSMutableDictionary dictionary];
    barTitleDic[NSForegroundColorAttributeName] = foregroundColor;
    [[UINavigationBar appearance] setTitleTextAttributes:barTitleDic];
}

- (void)setShadowImage:(UIImage *)shadowImage
{
    _shadowImage = shadowImage;
    self.navigationBar.shadowImage = shadowImage;
}

- (nullable UIImage *)bc_imageNamed:(NSString *)name{
    if(name &&
       ![name isEqualToString:@""]){
        NSBundle *bundle = [NSBundle bundleForClass:NSClassFromString(@"YKBaseViewController")];
        NSURL *url = [bundle URLForResource:@"YKOCBaseClass" withExtension:@"bundle"];
        if(!url) return [UIImage imageNamed:name]?:[UIImage new];
        NSBundle *imageBundle = [NSBundle bundleWithURL:url];
        
        NSString *imageName = nil;
        CGFloat scale = [UIScreen mainScreen].scale;
        if (ABS(scale-3) <= 0.001){
            imageName = [NSString stringWithFormat:@"%@@3x",name];
        }else if(ABS(scale-2) <= 0.001){
            imageName = [NSString stringWithFormat:@"%@@2x",name];
        }else{
            imageName = name;
        }
        UIImage *image = [UIImage imageWithContentsOfFile:[imageBundle pathForResource:imageName ofType:@"png"]];
        if (!image) {
            image = [UIImage imageWithContentsOfFile:[imageBundle pathForResource:name ofType:@"png"]];
            if (!image) {
                image = [UIImage imageNamed:name];
            }
        }
        return image;
    }
    
    return nil;
}

@end

File: Module/YKSwiftBaseClass/YKSwiftBaseClass/Classes/YKNavigationController.swift
//
//  YKNavigationController.swift
//  YKSwiftBaseClass
//
//  Created by linghit on 2021/11/17.
//

import UIKit


open class YKNavigationController : UINavigationController,UINavigationControllerDelegate
{
    
    public var backBtn:UIButton = UIButton(type: .custom)
    
    public var backBtnItem:UIBarButtonItem = UIBarButtonItem.init()
    
    private var _foregroundColor:UIColor = .white
    public var foregroundColor:UIColor{
        get{
            return _foregroundColor
        }
        set{
            _foregroundColor = newValue
            var barTitleDic = Dictionary<NSAttributedString.Key,Any>.init()
            barTitleDic[NSAttributedString.Key.foregroundColor] = newValue
            UINavigationBar.appearance().titleTextAttributes = barTitleDic
        }
    }
    
    private var _backgroundColor:UIColor = .white
    public var backgroundColor:UIColor{
        get{
            return _backgroundColor
        }
        set{
            _backgroundColor = newValue
            if #available(iOS 15.0, *) {
                let standardAppearance = navigationBar.standardAppearance
                standardAppearance.backgroundColor = newValue
                navigationBar.standardAppearance = standardAppearance
                
                let scrollEdgeAppearance = navigationBar.scrollEdgeAppearance ?? UINavigationBarAppearance.init()
                scrollEdgeAppearance.backgroundColor = newValue
                navigationBar.scrollEdgeAppearance = scrollEdgeAppearance
            } else {
                navigationBar.barTintColor = newValue
                navigationBar.backgroundColor = newValue
            }
        }
    }
    
    private var _shadowImage:UIImage = UIImage.init()
    public var shadowImage:UIImage{
        get{
            return _shadowImage
        }
        set{
            _shadowImage = newValue
            navigationBar.shadowImage = newValue
        }
    }
    
    private var _tabbarBackgroundColor:UIColor = .white
    public var tabbarBackgroundColor:UIColor{
        get{
            return _tabbarBackgroundColor
        }
        set{
            _tabbarBackgroundColor = newValue
            if #available(iOS 15, *) {
                let scrollAppearance = tabBarItem.scrollEdgeAppearance ?? UITabBarAppearance.init()
                scrollAppearance.backgroundColor = newValue
                tabBarItem.scrollEdgeAppearance = scrollAppearance
                
                let standardAppearance = tabBarItem.standardAppearance ?? UITabBarAppearance.init()
                standardAppearance.backgroundColor = newValue
                tabBarItem.standardAppearance = standardAppearance
                
            } else {
                UITabBar.appearance().tintColor = newValue
            }
        }
    }
    
    private var _tabbarNormalItemTextColor:UIColor = .white
    public var tabbarNormalItemTextColor:UIColor{
        get{
            return _tabbarNormalItemTextColor
        }
        set{
            _tabbarNormalItemTextColor = newValue
            if #available(iOS 15, *) {
                let scrollAppearance = tabBarItem.scrollEdgeAppearance ?? UITabBarAppearance.init()
                var scrollNormalAttributes = scrollAppearance.stackedLayoutAppearance.normal.titleTextAttributes
                scrollNormalAttributes[NSAttributedString.Key.foregroundColor] = newValue
                scrollAppearance.stackedLayoutAppearance.normal.titleTextAttributes = scrollNormalAttributes
                tabBarItem.scrollEdgeAppearance = scrollAppearance
                
                let standardAppearance = tabBarItem.standardAppearance ?? UITabBarAppearance.init()
                var standardNormalAttributes = standardAppearance.stackedLayoutAppearance.normal.titleTextAttributes
                standardNormalAttributes[NSAttributedString.Key.foregroundColor] = newValue
                standardAppearance.stackedLayoutAppearance.normal.titleTextAttributes = standardNormalAttributes
                tabBarItem.standardAppearance = standardAppearance
                
            } else {
                UITabBar.appearance().tintColor = newValue
                UITabBarItem.appearance().setTitleTextAttributes([NSAttributedString.Key.foregroundColor:newValue], for: .normal)
            }
        }
    }
    
    private var _tabbarSelectItemTextColor:UIColor = .white
    public var tabbarSelectItemTextColor:UIColor{
        get{
            return _tabbarSelectItemTextColor
        }
        set{
            _tabbarSelectItemTextColor = newValue
            if #available(iOS 15, *) {
                let scrollAppearance = tabBarItem.scrollEdgeAppearance ?? UITabBarAppearance.init()
                var scrollSelectedAttributes = scrollAppearance.stackedLayoutAppearance.selected.titleTextAttributes
                scrollSelectedAttributes[NSAttributedString.Key.foregroundColor] = newValue
                scrollAppearance.stackedLayoutAppearance.selected.titleTextAttributes = scrollSelectedAttributes
                tabBarItem.scrollEdgeAppearance = scrollAppearance
                
                let standardAppearance = tabBarItem.standardAppearance ?? UITabBarAppearance.init()
                var standardSelectedAttributes = standardAppearance.stackedLayoutAppearance.selected.titleTextAttributes
                standardSelectedAttributes[NSAttributedString.Key.foregroundColor] = newValue
                standardAppearance.stackedLayoutAppearance.selected.titleTextAttributes = standardSelectedAttributes
                tabBarItem.standardAppearance = standardAppearance
                
            } else {
                UITabBar.appearance().tintColor = newValue
                UITabBarItem.appearance().setTitleTextAttributes([NSAttributedString.Key.foregroundColor:newValue], for: .normal)
            }
        }
    }
    
    private var _tabbarSelectedImage:UIImage = UIImage.init()
    public var tabbarSelectedImage:UIImage{
        get{
            return _tabbarSelectedImage
        }
        set{
            _tabbarSelectedImage = newValue
            tabBarItem.selectedImage = newValue
        }
    }
    
    private var _tabbarNormalImage:UIImage = UIImage.init()
    public var tabbarNormalImage:UIImage{
        get{
            return _tabbarNormalImage
        }
        set{
            _tabbarNormalImage = newValue
            tabBarItem.image = newValue
        }
    }
    
    private var _tabbarTitle:String = ""
    public var tabbarTitle:String{
        get{
            return _tabbarTitle
        }
        set{
            _tabbarTitle = newValue
            tabBarItem.title = newValue
        }
    }
    
    private var _backBtnTitle:String? = nil
    public var backBtnTitle:String?{
        get{
            return _backBtnTitle
        }
        set{
            _backBtnTitle = newValue
            
            
            self.backBtn.setTitle(newValue, for: .normal)
            if #available(iOS 13.0, *) {
                self.backBtn.setTitleColor(.label, for: .normal)
            } else {
                self.backBtn.setTitleColor(.black, for: .normal)
            }
            let ipad = UIDevice.current.userInterfaceIdiom == .pad ? 2: 1;
            var frame = self.backBtn.frame
            frame.size = CGSize(width: 44 * ipad, height: 44 * ipad)
            self.backBtn.frame = frame
            self.backBtn.titleEdgeInsets = UIEdgeInsets.init(top: 0, left: 10, bottom: 0, right: 0)
            self.backBtn.contentHorizontalAlignment = .left
        }
    }
    
    public override init(rootViewController: UIViewController) {
        super.init(rootViewController: rootViewController)
        self.modalPresentationStyle = .fullScreen
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    open override func viewDidLoad() {
        super.viewDidLoad()
        self.delegate = self
    }
    
    open override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        
        var childrenVCs:[UIViewController] = []
        #if swift(>=4.2)
            childrenVCs = self.children
        #else
            childrenVCs = self.childViewControllers
        #endif
        
        if childrenVCs.count > 0 {
            let backBtn = UIButton.init(type: .custom)
            var url = ""
            let bundle = Bundle(for: YKNavigationController.self)
            if #available(iOS 13.0, *) {
                if traitCollection.userInterfaceStyle == .dark {
                    url = bundle.path(forResource: "ic_bc_back_while@3x", ofType: "png") ?? "ic_bc_back_while@3x.png"
                    
                }else{
                    url = bundle.path(forResource: "ic_bc_back_black@3x", ofType: "png") ?? "ic_bc_back_black@3x.png"
                }
                backBtn.setTitleColor(.label, for: .normal)
            } else {
                url = bundle.path(forResource: "ic_bc_back_black@3x", ofType: "png") ?? "ic_bc_back_black@3x.png"
                backBtn.setTitleColor(.black, for: .normal)
            }
            var backImage = UIImage.init(contentsOfFile: url)
            if backImage != nil {
                backImage = backImage!.withRenderingMode(.alwaysTemplate)
            }
            backBtn.setImage(backImage, for: .normal)
            backBtn.setImage(backImage, for: .highlighted)
            
            backBtn.titleLabel?.font = UIFont.systemFont(ofSize: 17)
            
            let ipad = UIDevice.current.userInterfaceIdiom == .pad ? 2: 1;
            var frame = backBtn.frame
            frame.size = CGSize(width: 44 * ipad, height: 44 * ipad)
            backBtn.frame = frame
            backBtn.contentHorizontalAlignment = .left
            backBtn.addTarget(self, action: #selector(back(sender:)), for: .touchUpInside)
            backBtn.sizeToFit()
            
            self.backBtn = backBtn
            
            viewController.navigationItem.leftBarButtonItem = UIBarButtonItem.init(customView: backBtn)
            
            
            self.backBtnItem = viewController.navigationItem.leftBarButtonItem!
            
            if viewControllers.count == 1 {
                viewController.hidesBottomBarWhenPushed = true
            }
            
        }else{
            viewController.hidesBottomBarWhenPushed = false
        }
        
        super.pushViewController(viewController, animated: animated)
    }
    
    @objc open func back(sender:AnyObject)->Void
    {
        self.popViewController(animated: true)
    }
    
    open override var preferredStatusBarStyle: UIStatusBarStyle
    {
        return .default
    }
}
